k8s/src/main/scala/io/buoyant/k8s/ServiceNamer.scala

```diff
@@ -0,0 +1,174 @@
+
+  private[this] val PrefixLen = 3
+
+  override def lookup(path: Path): Activity[NameTree[Name]] = path.take(PrefixLen) match {
+    case id@Path.Utf8(nsName, portName, serviceName) =>
+
+      val nameTree = serviceNs.get(nsName).get(serviceName, portName).map {
+        case Some(address) =>
+          val bound = address.map(Addr.Bound(_))
+          NameTree.Leaf(Name.Bound(bound, idPrefix ++ id, path.drop(PrefixLen)))
+        case None =>
+          NameTree.Neg
+      }
+      Activity(nameTree.map(Activity.Ok(_)))
+
+    case _ =>
+      Activity.value(NameTree.Neg)
+  }
+
+  private[this] def getPort(service: Service, portName: String): Option[Int] =
+    for {
+      spec <- service.spec
+      port <- spec.ports.find(_.name == portName)
+    } yield port.port
+
+  val serviceNs = new Ns[Service, ServiceWatch, ServiceList, ServiceCache](backoff, timer) {
+    override protected def mkResource(name: String): NsListResource[Service, ServiceWatch, ServiceList] =
+      mkApi(name).services
+
+    override protected def mkCache(name: String): ServiceCache =
+      new ServiceCache(name)
+
+    override protected def update(cache: ServiceCache)(event: ServiceWatch): Unit =
+      cache.update(event)
+
+    override protected def initialize(
+      cache: ServiceCache,
+      list: ServiceList
+    ): Unit = cache.initialize(list)
+  }
+}
+
+class ServiceCache(namespace: String) {
+
+  def get(serviceName: String, portName: String): Var[Option[Var[Address]]] = synchronized {
+    // we call this unstable because every change to the Address will cause
+    // the entire Var[Option[Address]] to update.
+    val unstable = cache.get(serviceName) match {
+      case Some(ports) => ports.map(_.get(portName))
+      case None =>
+        val ports = Var(Map.empty[String, Address])
+        cache += serviceName -> ports
+        ports.map(_.get(portName))
+    }
+
+    // We can stabilize this by changing the type to Var[Option[Var[Address]]].
+    // If this service port is created or deleted, the outer Var will update.
+    // If the address of the service port changes, only the inner Var will
+    // update.
+    val init = unstable.sample().map(Var(_))
+    Var.async[Option[VarUp[Address]]](init) { update =>
+      // the current inner Var, null if the outer Var is None
+      @volatile var addr: VarUp[Address] = null
+
+      unstable.changes.respond {
+        case Some(address) if addr == null =>
+          // Address created
+          addr = Var(address)
+          update() = Some(addr)
+        case Some(address) =>
+          // Address modified
+          addr() = address
+        case None =>
+          // Address deleted
+          addr = null
+          update() = None
+      }
+    }
+  }
+
+  private[this]type VarUp[T] = Var[T] with Updatable[T]
+
+  private[this] var cache = Map.empty[String, VarUp[Map[String, Address]]]
+
+  private[this] def extractPorts(service: Service): Map[String, Address] = {
```

Comment:
tioli: there are a few stateless private helpers in this file.  You could move them into a companion object to more clearly differentiate static-helper and stateful-updatey logic